sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -238,8 +238,9 @@ class Dataset[T] private[sql](
   private[sql] def showString(
       _numRows: Int, truncate: Int = 20, vertical: Boolean = false): String = {
     val numRows = _numRows.max(0)
-    val takeResult = toDF().take(numRows + 1)
-    val hasMoreData = takeResult.length > numRows
+    val numTotalRows = toDF().count()
```

Comment:
You don't want to do a whole count() here -- could be quite expensive. Instead just something like:

```
    val takeResult = toDF().take(if (numRows == Int.MaxValue) numRows else numRows + 1)
    val hasMoreData = takeResult.length > numRows
    val data = takeResult.take(numRows)
```

Reply:
ok, I'll update

Reply:
In the suggested, `hasMoreData` gets meaningless, so how about this?;
```
    val (data, hasMoreData) = if (numRows < Int.MaxValue) {
      val takeResult = toDF().take(numRows + 1)
      (takeResult.take(numRows), takeResult.length > numRows)
    } else {
      val takeResult = toDF().take(numRows)
      val numTotalRows = toDF().count()
      (takeResult.take(numRows), numTotalRows > numRows)
    }
```